File: core/assert/string_test.go
// Copyright (C) 2017 Google Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package assert_test

import "github.com/google/gapid/core/assert"

// An example of testing string equality
func ExampleStringEquals() {
	assert := assert.To(nil)
	assert.For(`"a" Equals "a"`).ThatString("a").Equals("a")
	assert.For(`"a" Equals "b"`).ThatString("a").Equals("b")
	assert.For(`"a" NotEquals "a"`).ThatString("a").NotEquals("a")
	assert.For(`"a" NotEquals "b"`).ThatString("a").NotEquals("b")
	// Output:
	// Error:"a" Equals "b"
	//     Got          `a`
	//     Expect  ==   `b`
	//     Differs from `a`
	// Error:"a" NotEquals "a"
	//     Got       `a`
	//     Expect != `a`
}

// An example of testing partial string equality
func ExampleStringFragments() {
	assert := assert.To(nil)
	assert.For(`"abc" Contains "b"`).ThatString("abc").Contains("b")
	assert.For(`"abc" Contains "d"`).ThatString("abc").Contains("d")
	assert.For(`"abc" HasPrefix "a"`).ThatString("abc").HasPrefix("a")
	assert.For(`"abc" HasPrefix "b"`).ThatString("abc").HasPrefix("b")
	assert.For(`"abc" HasSuffix "c"`).ThatString("abc").HasSuffix("c")
	assert.For(`"abc" HasSuffix "b"`).ThatString("abc").HasSuffix("b")
	// Output:
	// Error:"abc" Contains "d"
	//     Got             `abc`
	//     Expect contains `d`
	// Error:"abc" HasPrefix "b"
	//     Got                `abc`
	//     Expect starts with `b`
	// Error:"abc" HasSuffix "b"
	//     Got              `abc`
	//     Expect ends with `b`
}

// An example of testing non strings as strings
func ExampleStringTypes() {
	assert := assert.To(nil)
	assert.For(`"abc" Equals "a"`).ThatString([]byte{'a', 'b', 'c'}).Equals("a")
	assert.For(`10 Equals "10"`).ThatString(10).Equals("10")
	// Output:
	// Error:"abc" Equals "a"
	//     Got       `abc`
	//     Expect == `a`
	//     Longer by `bc`
}

File: gapil/runtime/cc/string_test.cpp
// Copyright (C) 2018 Google Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "string.h"

#include "core/memory/arena/cc/arena.h"

#include <gtest/gtest.h>

TEST(StringTest, empty) {
  gapil::String str;

  EXPECT_EQ(str.length(), 0);

  EXPECT_EQ(str, gapil::String());

  EXPECT_STREQ(str.c_str(), "");
}

TEST(StringTest, constructors) {
  auto abc123 = "abc123";
  core::Arena arena;
  {
    gapil::String str;
    EXPECT_EQ(str.length(), 0);
    EXPECT_STREQ(str.c_str(), "");
  }
  {
    gapil::String strA(&arena, abc123);
    gapil::String strB(strA);
    EXPECT_EQ(strB.length(), 6);
    EXPECT_STREQ(strB.c_str(), abc123);
  }
  {
    gapil::String strA(&arena, abc123);
    gapil::String strB(std::move(strA));
    EXPECT_EQ(strB.length(), 6);
    EXPECT_STREQ(strB.c_str(), abc123);
  }
  {
    gapil::String str(&arena, abc123);
    EXPECT_EQ(str.length(), 6);
    EXPECT_STREQ(str.c_str(), abc123);
  }
  {
    gapil::String str(&arena, {'a', 'b', 'c'});
    EXPECT_EQ(str.length(), 3);
    EXPECT_STREQ(str.c_str(), "abc");
  }
  {
    gapil::String str(&arena, &abc123[2], &abc123[4]);
    EXPECT_EQ(str.length(), 2);
    EXPECT_STREQ(str.c_str(), "c1");
  }
  {
    gapil::String str(&arena, &abc123[2], 3);
    EXPECT_EQ(str.length(), 3);
    EXPECT_STREQ(str.c_str(), "c12");
  }

  EXPECT_EQ(arena.num_allocations(), 0);  // nothing leaked
}

TEST(StringTest, allocs) {
  core::Arena arena;

  {
    gapil::String str(&arena, "hello world");

    EXPECT_EQ(arena.num_allocations(), 1);  // str owns 1 allocation

    EXPECT_EQ(str.length(), 11);

    EXPECT_EQ(str, gapil::String(&arena, "hello world"));

    EXPECT_EQ(arena.num_allocations(), 1);  // temporary has been freed

    EXPECT_STREQ(str.c_str(), "hello world");
  }

  EXPECT_EQ(arena.num_allocations(), 0);  // str has fallen out of scope
}

TEST(StringTest, assignment) {
  core::Arena arena;

  {
    gapil::String strA(&arena, "hello world");

    EXPECT_EQ(arena.num_allocations(), 1);  // strA owns 1 allocation

    gapil::String strB;

    EXPECT_EQ(arena.num_allocations(), 1);  // strB has made no allocations

    strB = strA;

    EXPECT_EQ(arena.num_allocations(), 1);  // strB is sharing strA's allocation

    strA = gapil::String();

    EXPECT_EQ(arena.num_allocations(),
              1);  // strB now exclusively owns the allocation

    strA = strB;

    EXPECT_EQ(arena.num_allocations(), 1);

    EXPECT_STREQ(strB.c_str(), "hello world");
  }

  EXPECT_EQ(arena.num_allocations(), 0);
}

TEST(StringTest, compare) {
  core::Arena arena;

  gapil::String strA(&arena, "meow");
  gapil::String strB(&arena, "woof");

  EXPECT_FALSE(strA == strB);
  EXPECT_TRUE(strA != strB);
  EXPECT_TRUE(strA < strB);
  EXPECT_TRUE(strA <= strB);
  EXPECT_TRUE(strB > strA);
  EXPECT_TRUE(strB >= strA);
}

TEST(StringTest, clear) {
  core::Arena arena;

  gapil::String str(&arena, "hello");

  EXPECT_EQ(arena.num_allocations(), 1);

  str.clear();

  EXPECT_EQ(arena.num_allocations(), 0);
  EXPECT_STREQ(str.c_str(), "");
}
